cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -354,13 +377,397 @@ public KafkaAssemblyOperator(Vertx vertx, boolean isOpenShift,
+                    }
+                    log.debug("Does SS {} need to be upgraded?", ss.getMetadata().getName());
+                    Future<?> result;
+                    // Get the current version of the cluster
+                    KafkaVersion currentVersion = versions.version(ss.getMetadata().getAnnotations().get(KafkaCluster.ANNO_STRIMZI_IO_KAFKA_VERSION));
+                    log.debug("SS {} has current version {}", ss.getMetadata().getName(), currentVersion);
+                    String fromVersionAnno = ss.getMetadata().getAnnotations().get(ANNO_STRIMZI_IO_FROM_VERSION);
+                    KafkaVersion fromVersion;
+                    if (fromVersionAnno != null) { // We're mid-upgrade
+                        fromVersion = versions.version(fromVersionAnno);
+                    } else {
+                        fromVersion = currentVersion;
+                    }
+                    log.debug("SS {} is upgrading from version {}", ss.getMetadata().getName(), fromVersion);
+                    String toVersionAnno = ss.getMetadata().getAnnotations().get(ANNO_STRIMZI_IO_TO_VERSION);
+                    KafkaVersion toVersion;
+                    if (toVersionAnno != null) { // We're mid-upgrade
+                        toVersion = versions.version(toVersionAnno);
+                    } else {
+                        toVersion = versions.version(kafkaAssembly.getSpec().getKafka().getVersion());
+                    }
+                    log.debug("SS {} is upgrading to version {}", ss.getMetadata().getName(), toVersion);
+                    KafkaUpgrade upgrade = new KafkaUpgrade(fromVersion, toVersion);
+                    log.debug("Kafka upgrade {}", upgrade);
+                    if (upgrade.isNoop()) {
+                        log.debug("Kafka.spec.kafka.version unchanged");
+                        result = Future.succeededFuture();
+                    } else {
+                        String image = imageForVersion(toVersion);
+                        Future<StatefulSet> f = Future.succeededFuture(ss);
+                        if (upgrade.isUpgrade()) {
+                            if (currentVersion.equals(fromVersion)) {
+                                f = f.compose(ignored -> kafkaUpgradePhase1(ss, upgrade, image));
+                            }
+                            result = f.compose(ss2 -> kafkaUpgradePhase2(ss2, upgrade));
+                        } else {
+                            if (currentVersion.equals(fromVersion)) {
+                                f = f.compose(ignored -> kafkaDowngradePhase1(ss, upgrade));
+                            }
+                            result = f.compose(ignored -> kafkaDowngradePhase2(ss, upgrade, image));
+                        }
+
+                    }
+                    return result.map(this);
+                });
+        }
+
+        private String imageForVersion(KafkaVersion toVersion) {
```

Comment:
Wouldn't this be better in the Model classes or in some Utils?

Reply:
The model doesn't have the necessary info (`Kafka` instance or image map), so in both cases any method would need to have parameters for those. And right now there is just this one caller, so I'm inclined to leave it where it is for now.